Create shell scripts for installing, running, and testing application

coltonlw opened the pull request:
Resolves #419 

- Adds and modifies scripts in bin/ for non-docker way to install and test application, on ubuntu and osx
- Adds scripts in test/bin/ for testing the application outside of docker
- Disables insecure mode for integration tests and runs them using user API key
- Changes bootstrap.py to load users directly into the database, allowing an admin to add a user with an API key
- Updates README.md and TESTING.md to document these changes


**Discussed Actions before merge 2016-09-02**
- [ ] Colton Bring back support for env vars overriding env config file
- [ ] Gunnar Try again at installing and using node inside virtual env
- [ ] Gunnar to submit other PR for desired tweaks
- [ ] Ryan make change for repo consumer to remove pyc files in Integration tests

coltonlw commented:
Closing while I fix coveralls

coltonlw commented:
Coveralls says the result was sent and the link in github goes to the completed result, so re-opening

gsfr commented:
Feel free to leave your PRs open, even if they are not quite ready to be merged. I find the repeated closing and re-opening notifications rather confusing. Our usual flow is that the PR author gets to close the PR then they feel that it's ready, after getting thumbs up from the reviewers.

coltonlw commented:
I see, I didn't read the last sentence carefully enough I'll just leave
them open from now on

On Mon, Aug 29, 2016 at 7:26 PM, Gunnar Schaefer <notifications@github.com>
wrote:

> Feel free to leave your PRs open, even if they are not quite ready to be
> merged. I find the repeated closing and re-opening notifications rather
> confusing. Our usual flow is that the PR author gets to close the PR then
> they feel that it's ready, after getting thumbs up from the reviewers.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/scitran/core/pull/442#issuecomment-243298020>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AT8Lk5MFxDeD3VAR6e_548kf9D3EJBpVks5qk3i1gaJpZM4JvwjJ>
> .
>



-- 
Colton Leekley-Winslow
*Backend Developer*
Flywheel.io <https://flywheel.io/>

coltonlw commented:
@ryansanford @gsfr I moved the new bootstrap script for loading users into the database to "bootstrap-dev.py"  until we can complete the separate task of finalizing this into a "create_superuser" command which generates the one users api key and returns it.  I also added the old bootstrap.py in bin/load_users_drone_secret.py

gsfr commented on CONTRIBUTING.md:
```diff
@@ -23,7 +23,7 @@ Changes to `requirements.txt` should always be by pull request.
 
 ### Format
-- Ensure that `./test/lint.sh api` exits without errors.
+Ensure that `./test/bin/lint.sh api` exits without errors.
```
The single bullet is silly, but I would keep it for symmetry.

gsfr commented on bin/api.wsgi:
```diff
@@ -1,4 +1,9 @@
+import os.path
+
+repo_path = os.path.abspath(os.path.join(os.path.dirname(__file__), '../'))
+sys.path.append(repo_path)
```
As far as I'm concerned, any kind of internal, *magic* manipulation of the PYTHONPATH is evil and should be avoided.

gsfr replied:
Also, please undo the file mode change.

coltonlw replied:
Fixed both of these

gsfr commented on bin/bootstrap-dev.py:
```diff
@@ -0,0 +1,51 @@
+import jsonschema
+
+repo_path = os.path.abspath(os.path.join(os.path.dirname(__file__), '../'))
+sys.path.append(repo_path)
```
Same here.

gsfr commented on bin/bootstrap-dev.py:
```diff
@@ -0,0 +1,51 @@
+repo_path = os.path.abspath(os.path.join(os.path.dirname(__file__), '../'))
+sys.path.append(repo_path)
+
+from api import config, validators
```
For clarity, I prefer `import api` and then use of `api.config`, etc.

coltonlw replied:
Do you mean `import api.config`?  When importing a package, modules / sub-packages are not imported and attached unless you specify that in `__init__.py`

gsfr replied:
Yes, my bad.

We could also update `__init__.py` to auto-import the usual suspects.

gsfr commented on bin/run-dev-osx.sh:
```diff
@@ -0,0 +1,165 @@
+  shift
+done
+
+set -a
```
Let's leave this at `set -o allexport` so that I don't have to look up what `-a` is. Also, you are missing the `+a`.

gsfr commented on requirements.txt:
```diff
@@ -10,6 +10,7 @@ requests==2.9.1
 requests-toolbelt==0.6.0
 rfc3987==1.3.4
 strict-rfc3339==0.7
+uwsgi==2.0.13.1
```
Does this dependency also get used in prod?

ryansanford commented:
@coltonlw 
My changes are pushed.

gsfr commented on bin/install-osx-dev.sh:
```diff
@@ -80,7 +80,7 @@ fi
 
 # Install testing dependencies
-pip install -r "test/integration_tests/requirements.txt"
+pip install -r "test/integration_tests/requirements-integration-test.txt"
```
@ryansanford Do we really need "integration_test" twice in the path? Doesn't that defeat the purpose of subdirectories?

gsfr commented:
@coltonlw 

In `run-dev-osx.sh`
- `AUTO_RELOAD_INTERVAL` is not used, but should be
- `CONFIG_FILE` is not sourced, but should be
- SSL support needs to be restored
- `echo` highlighting needs to be restored

ryansanford commented:
@gsfr I'm not tied to the exact suffix, but it should not be just `requirements.txt` as that makes finding references in the project ambiguous. We established that naming practice around February of this year.

coltonlw commented:
@gsfr RE your last comment:

- Fixed usage of AUTO_RELOAD_INTERVAL
- Fixed sourcing CONFIG_FILE
- SSL is supported by passing a uwsgi ini file
- Can you help me with echo highlighting?  This is what I see:

```
(scitran) Coltons-MacBook-Pro:core coltonlw$ echo() { builtin echo -e "\e[1;7mSCITRAN\e[0;7m $@\e[27m"; }
(scitran) Coltons-MacBook-Pro:core coltonlw$ echo test
\e[1;7mSCITRAN\e[0;7m test\e[27m
(scitran) Coltons-MacBook-Pro:core coltonlw$
```

coltonlw commented:
@gsfr had a couple more things to clean up after I replied to your comment, the branch is stable now

gsfr commented:
Thanks, @coltonlw. I'll check out all of your updates later today.

Re SSL, please restore support for the `SCITRAN_RUNTIME_SSL_PEM` environment variable.

Re echo highlighting, my guess is that you are using an ancient version of bash. macOS ships with 3.2, but 4.3 is current. Of course it would be great to have this work in the old bash, too, but let's not worry about that.

Here is how I install and use the latest bash via homebrew:
```
brew install bash bash-completion
sudo dscl . -create /Users/$(whoami) UserShell /usr/local/bin/bash
```

ryansanford commented:
@gsfr @coltonlw If the run dev osx script assumes a non-stock version of bash on OSX, please indicate that in the README.md next to the launch instructions.

coltonlw commented:
@ryansanford +1, got that in there

coltonlw commented:
@gsfr Echo highlighting and SSL restored

gsfr commented:
I hope this is ready to go, despite the triple revert. 😉 

LGTM, although I was thinking we'd keep the note about upgrading bash, after all. Either way, please merge at your leisure.
